### ISSUE TITLE
[Performance/RegexpMatch] Introducing bug on `foo =~ /re/` to `foo.match?(/re/)`

### ISSUE THREAD

**Issue by Dorian** [bbatsov/rubocop#4298]
Transforming `foo =~ /re/` to `foo.match?(/re/)` fails when foo is `nil`. Thus introducing a bug in the codebase.

--------

## Expected behavior

- Not doing anything on variables?
- Some kind of warning?
- `foo.to_s.match?(/re/)`?

## Actual behavior

- `foo.match?(/re/)` fails on `foo = nil`

## Steps to reproduce the problem

There is even a spec for it :) https://github.com/bbatsov/rubocop/blob/master/spec/rubocop/cop/performance/regexp_match_spec.rb#L290

```
['matching by =~`', 'foo =~ /re/', 'foo.match?(/re/)']
```

## RuboCop version

Weirdly enough, I can't reproduce on my machine but my coworker could.

```
> rubocop -V
0.48.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-darwin16)
```

**Comment by pocke** [comment]
I work on it. 

> Not doing anything on variables?


I think this solution is better.

**Comment by Drenmi** [comment]
> Thus introducing a bug in the codebase.

Depends how one sees it, I guess. To me, this looks like it's *revealing* a bug. Comparing `nil` to anything (except maybe to `nil`) is meaningless. The problem is likely `foo` being `nil` in the first place, which was masked by the liberal treatment in `=~`. 🙂 

That said, we should be conservative that auto-correct doesn't break existing code, so we should probably not correct in these cases. I still think we should report an offense.

**Comment by pocke** [comment]
@Drenmi Thanks for your suggestion. Sounds good to me.

>  so we should probably not correct in these cases. I still think we should report an offense.

:+1:

**Comment by Dorian** [comment]
Thanks

**Comment by jhottenstein** [comment]
For our codebase, I autocorrected to =~  to &.match?

**Comment by rrosenblum** [comment]
@pocke any updates on this? I just ran into this issue and was going to work on it until I saw this existing issue. My plan was to correct variables using `=~` to use `&.match?`. String, symbol, and regex literals can continue to use `.match?`.

**Comment by pocke** [comment]
I do not work on this issue now. Thank you.

> My plan was to correct variables using `=~` to use `&.match?`. String, symbol, and regex literals can continue to use `.match?`.

I think replacing `=~` with `&.match?` is meaningful only in limited cases. When `a` is always non nil, `a&.match?` is meaningless. So I think it is not the best way.
Probably we can swap the receiver and the parameter when the receiver is a variable.

```ruby
# correct to /a/.match?(var)
var =~ /a/ 

# Probably we can auto-correct it to SOME_REGEXP.match?(var)
# Because constant is not a nil in many cases.
var =~ SOME_REGEXP

# In my opinion, I think the cop should not auto-correct the code.
# `var.match?(var2)` will make a bug, and `var&.match?(var2)` will be meaningless in other cases.
var =~ var2
```

**Comment by rrosenblum** [comment]
I like the idea of swapping the receiver and the parameter if there is a constant, string literal, or regexp literal used as the parameter.

> When a is always non nil, a&.match? is meaningless.

True, however we have no way of knowing if `a` is always going to be non `nil`. I'm not sure what the aversion to using safe navigation for the auto-correction of two variables is. I understand that it is meaningless when the variable will never be `nil`. It is a legitimate way of safe guarding against a `NoMethodError`, and it is a clean and simple solution.

**Comment by pocke** [comment]
>  I'm not sure what the aversion to using safe navigation for the auto-correction of two variables is.

I tend to use `=~` without conscious of ignoring `nil` in almost cases, so I guess I will be surprised by replacing it with `&.`.

However, `=~` and `&.match?` probably have same meaning. So I think the replacing is correct way.

I would like to know other people feel it surprising or not. If many people feel it is surprising, we should not auto-correct this case IMO.

**Comment by Drenmi** [comment]
I think the lonely operator is by far the worst new feature, so naturally not very happy to see it introduced by auto-correct. 😅
